libp2p/switch.nim

```diff
@@ -574,44 +430,27 @@ proc subscribePeer*(s: Switch, peerInfo: PeerInfo) {.async, gcsafe.} =
       if not(isNil(stream)):
         await stream.close()
 
-    if isNil(stream):
-      trace "unable to subscribe to peer", peer = peerInfo.shortLog
-      return
-
-    if not await s.ms.select(stream, s.pubSub.get().codec):
-      if not(isNil(stream)):
-        await stream.close()
-      return
-
-    s.pubSub.get().subscribePeer(stream)
-
 proc subscribe*(s: Switch, topic: string,
-                handler: TopicHandler): Future[void] =
+                handler: TopicHandler) {.async.} =
   ## subscribe to a pubsub topic
   if s.pubSub.isNone:
-    var retFuture = newFuture[void]("Switch.subscribe")
-    retFuture.fail(newNoPubSubException())
-    return retFuture
+    raise newNoPubSubException()
 
-  return s.pubSub.get().subscribe(topic, handler)
+  await s.pubSub.get().subscribe(topic, handler)
 
-proc unsubscribe*(s: Switch, topics: seq[TopicPair]): Future[void] =
+proc unsubscribe*(s: Switch, topics: seq[TopicPair]) {.async.} =
   ## unsubscribe from topics
   if s.pubSub.isNone:
-    var retFuture = newFuture[void]("Switch.unsubscribe")
-    retFuture.fail(newNoPubSubException())
-    return retFuture
+    raise newNoPubSubException()
 
-  return s.pubSub.get().unsubscribe(topics)
+  await s.pubSub.get().unsubscribe(topics)
 
-proc publish*(s: Switch, topic: string, data: seq[byte]): Future[int] =
+proc publish*(s: Switch, topic: string, data: seq[byte]): Future[int] {.async.} =
   # pubslish to pubsub topic
   if s.pubSub.isNone:
-    var retFuture = newFuture[int]("Switch.publish")
-    retFuture.fail(newNoPubSubException())
-    return retFuture
+    raise newNoPubSubException()
 
-  return s.pubSub.get().publish(topic, data)
+  return await s.pubSub.get().publish(topic, data)
```

Comment:
Transformation not required still

Reply:
it is, otherwise you can't use `raise` and have to return a manually constructed `Future`, this is cleaner with very little overhead if any.